### PR TITLE
Quarkus3 upgrade

### PR DESCRIPTION
**Pull request by mabartos**
It's possible to build a module `model` with Hibernate 6 and JPA 3; all changes made by @sguilhen there.

It's also possible to build Quarkus 3 distribution now.

However, I've encountered an issue with:

- Quarkus vault - Unsatisfied deps (vault ignored for now) -> issue with the Default Bean within CDI.

### CODE REVIEW

**Comment by pedroigor** [comment]
@mabartos Does it mean we need a specific ISPN version that is using Jakarta APIs?

**Comment by mabartos** [comment]
> @mabartos Does it mean we need a specific ISPN version that is using Jakarta APIs?

ISPN release two versions of their deps ATM. The classic one and one with `-jakarta` suffix. You can see it here[1]. When I'm trying to run the server, during the initialization of caches, there's thrown an exception in some ISPN library regarding fact, that it's necessary to use `javax`. So, there will be some problem with transitive deps for some library. From log, it seems it's related to HotRod.

However, HotRod shouldn't be present by default in Keycloak, right?

[1] https://github.com/keycloak/keycloak/pull/16262/commits/d385930faab94649e63651a7cdc0be3df8ed9321#diff-af770532ae57da77677ee7f8aa88730e6af4dfe7aac0bda7a8843a8646b696feR37

**Comment by pedroigor** [comment]
The library is there and it might happen that during bootstrap some code is referencing HotRod (for instance, during config?) even though not used. Better ask Michal.

**Comment by sguilhen** [comment]
To get it running with undertow-embedded, I had to add a few more exclusions to some infinispan libs that were pulling the infinispan-commons with javax packages:

https://github.com/keycloak/keycloak/commit/2b3c496dace656e8c6fb5670e28ea2b47bd6b5df

**Comment by mabartos** [comment]
@pedroigor Yep, the HotRod is probably used only for the possible configuration of HotRod, but it's not used, right @mhajas ? I believe it's not gonna be such a problem with finding the cause for the deps inconsistencies.

**Comment by sguilhen** [comment]
@mabartos Tried to build this PR locally, failed with
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:testCompile (default-testCompile) on project db-allocator-plugin: Compilation failure
[ERROR] /home/sguilhen/Workspace/workspace-keycloak/keycloak/testsuite/db-allocator-plugin/src/test/java/org/keycloak/testsuite/dballocator/client/DBAllocatorServiceClientTest.java:[3,29] package org.apache.commons.io does not exist
```

**Comment by mabartos** [comment]
@sguilhen @mhajas Thanks for your info.

Now, it's possible to build all modules in Keycloak and start Keycloak on Quarkus 3.x! :tada: :tada:

image 

image

The only thing is that it still uses Hibernate 5 (probably because of a Quarkus Hibernate ORM extensions).

But I think this is a good base for the following work! :tada:

**Comment by sguilhen** [comment]
@mabartos This is great. I have a couple of commits that I will be sending to my branch, feel free to cherry-pick them into your commit.

In one of them I've re-enabled the undertow adapter modules so we can run KeycloakServer and the testsuite with undertow.

One comment about this PR: I would not make any changes to the disabled adapter modules. They target containers that use javax instead of jakarta and since you are disabling their modules I would not personally touch them.

**Comment by sguilhen** [comment]
@mabartos See top two commitsin https://github.com/sguilhen/keycloak/commits/quarkus3-test

They are using this PR as a base. One of them is re-enabling the undertow adapter modules, which are needed if we want to run KeycloakServer or the any tests using undertow-embedded. The other commit contains initial fixes for the JPA Map Storage.

**Comment by mabartos** [comment]
@sguilhen You're right about the adapters. I'd suggest keeping it right now instead of amending or adding new commits related to that in order to avoid some inconsistencies. We can polish this later.

Thanks! I'll cherry-pick your commits.

**Comment by mabartos** [comment]
ATM, it's not possible to run Quarkus 3 and Hibernate 6 as it requires additional changes for `quarkus-hibernate-orm` extension from the Quarkus team.

**Comment by mabartos** [comment]
@sguilhen I've added one commit regarding the adapters. The Undertow Jakarta EE adapter is automatically built only in the testsuite, so the Keycloak Undertow adapter is unaffected. It's necessary to also build an Undertow SAML adapter with the JakartaEE.

**Comment by ahus1** [comment]
**TL;DR:** Review ok from the commits I reviewed, Hynek will review the remaining storage commits.

I reviewed the commits "Model upgrade Hibernate/JPA" and "Map Storage JPA fixes" added by @sguilhen, and they look good to me. For all commented out code blocks / hacks I found in those commits there is an open issue on the storage team's board.

It is a known issue that Quarkus 3.alpha-2 still uses Hibernate 5.6 "Jakarta" flavor. Therefore, when running Quarkus in this branch with JPA Map storage, this fails with an exception "org.hibernate.annotations.Type missing element type". This is expected. At least the startup of Quarkus 3 with the legacy store seems to work. For any Hibernate 6 tests, use the Undertow setup.

Let's hope for Hibernate 6 to make it into Quarkus 3 soon, and hope that any other upcoming fixes in the store for Hibernate 6 break the Quarkus 3 + Hibernate 5 setup.

**Review by ahus1 — APPROVED** [review]

**Comment by mabartos** [comment]
@ahus1 Thanks for approving!

> It is a known issue that Quarkus 3.alpha-2 still uses Hibernate 5.6 "Jakarta" flavor. Therefore, when running Quarkus in this branch with JPA Map storage, this fails with an exception "org.hibernate.annotations.Type missing element type". This is expected. At least the startup of Quarkus 3 with the legacy store seems to work. For any Hibernate 6 tests, use the Undertow setup.

You're right. We need to wait for the Quarkus upgrade of `quarkus-hibernate-orm` extension with the support of Hibernate 6. I don't understand one thing; AFAIK, the Legacy JPA store was also upgraded in terms of using Hibernate 6, so how is possible that Quarkus is running correctly with the legacy store?

**Comment by mabartos** [comment]
For the proper execution of tests, we need to solve this issue[1] (thanks, @sguilhen for the PR!).

As the build phase of this PR is successful, I'd consider this PR as mergeable.

@hmlnarik @vmuzikar @pedroigor Could you please briefly check it? Thanks!

[1] https://github.com/arquillian/arquillian-container-undertow/pull/50

**Comment by ahus1** [comment]
@mabartos 

> the Legacy JPA store was also upgraded in terms of using Hibernate 6, so how is possible that Quarkus is running correctly with the legacy store? AFAIK, the Quarkus server should use JPA and H2 DB by default, right?

What we see here is that for compilation and for Undertow Hibernate 6 is used, while Quarkus deployment resolves Hibernate 5 Jakarta. While the first part is our choice here, the second part is expected. 

Starting Quarkus with JPA legacy store will still work as long as we use functionality that works both in Hibernate 5 (Jakarata) and Hibernate 6. This is also the default setup as you mentioned. 

For the newer JPA map store, we are already using some functionality that is only available in Hibernate 6 as part of the migration to fix something that was broken in Undertow + Hibernate 6, therefore the newer JPA map store will not work in Quarkus 3 until a new release of Quarkus 3 provides Hibernate 6. 

Going forward, we (the store team) should try to not use Hibernate 6 only features in the legacy store so we don't break too much existing functionality. Still I don't know yet if this is possible. Time will tell. If that happens, we need to re-assess things; there are still several other map stores that should work. I tested the concurrent hash map (`bin/kc.sh start-dev --storage chm`) and the startup works fine.

As a plan B, there could be a branch with Hibernate 5 Jakarata + Quarkus 3, and another branch with Hibernate 6 + Undertow. As a downside this would involve more overhead in handling branches. 

Talking to @hmlnarik yesterday we want to continue on this route for now and keep things together in one branch as this PR suggests.

**Comment by hmlnarik** [comment]
The commits assigned to me are reviewed and good to go in this first round. Thanks @sguilhen and @mabartos!

**Comment by mabartos** [comment]
@ahus1 Thanks for the info! Very informative!

> Starting Quarkus with JPA legacy store will still work as long as we use functionality that works both in Hibernate 5 (Jakarata) and Hibernate 6. This is also the default setup as you mentioned.

Nice!
------------------------------

@hmlnarik Thank you!

File: adapters/spi/undertow-adapter-spi/src/main/java/org/keycloak/adapters/undertow/UndertowHttpFacade.java [hunk 1]
```diff
@@ -196,18 +196,7 @@ public InputStream getInputStream(boolean buffered) {
...
-                        inputStream.mark(0);
-                        return new ServletInputStream() {
-                            @Override
-                            public int read() throws IOException {
-                                return inputStream.read();
-                            }
-                        };
-                    }
-                });
+                context.setServletRequest(UndertowHttpServletRequest.setupServletInputStream(servletRequest, inputStream));
```
Comment (pedroigor) [inline adapters/spi/undertow-adapter-spi/src/main/java/org/keycloak/adapters/undertow/UndertowHttpFacade.java:199]: This might cause some tests to fail because some adapters allow reading the request input twice. For instance, the policy enforcer relies on this when processing claim information points if the application is also reading the input stream.

But we should be able to get failures if that happen.

**Review by pedroigor — APPROVED** [review]

File: adapters/spi/undertow-adapter-spi/src/main/java/org/keycloak/adapters/undertow/UndertowHttpFacade.java (hunk 1 above)
Comment (mabartos) [reply adapters/spi/undertow-adapter-spi/src/main/java/org/keycloak/adapters/undertow/UndertowHttpFacade.java:199]: @pedroigor Thanks!
The static method in the class `UndertowHttpServletRequest` is only a wrapper of the old behavior, so IMHO, the logic behind it didn't change. Or did it?

The purpose of this change is to be able to override the class in the testsuite for Undertow Jakarta EE as it requires some other props for the servlet request.
